File: 0x08-python-more_classes/2-rectangle.py
#!/usr/bin/python3
"""Defining class Rectangle."""


class Rectangle:
    """This is an empty class called Rectangle."""

    def __init__(self, width=0, height=0):
        """Initialize a new Rectangle instance.

        Args:
            width (int): The width of the  rectangle.
            height (int): The height of the rectangle.


        """
        self.width = width
        self.height = height

    @property
    def width(self):
        """Get the width of the Rectangle."""
        return self._width

    @width.setter
    def width(self, value):
        """Set the width of the Rectangle."""
        if not isinstance(value, int):
            raise TypeError("width must be an integer")
        elif value < 0:
            raise ValueError("width must be >= 0")
        self._width = value

    @property
    def height(self):
        """Get the height of the Rectangle."""
        return self._height

    @height.setter
    def height(self, value):
        """Set the height of the Rectangle."""
        if not isinstance(value, int):
            raise TypeError("height must be an integer")
        elif value < 0:
            raise ValueError("height must be >= 0")
        self._height = value

    def area(self):
        """Calculate and return the area of a rectangle."""
        return self.width * self.height

    def perimeter(self):
        """Calculate and return the perimeter of a rectangle."""
        return 2 * (self.width + self.height)

File: 0x13-javascript_objects_scopes_closures/2-rectangle.js
#!/usr/bin/node
/*
  This class represents a Rectangle.

  It has a constructor that takes two arguments:
  - 'w' for width
  - 'h' for height

  The constructor initializes the instance attributes 'width' and 'height'
  with the values of 'w' and 'h', respectively.
*/
class Rectangle {
  constructor (w, h) {
    /*
      Inside the constructor:
      - 'this' refers to the newly created instance of the Rectangle class.
      - 'this.width' is an instance attribute that represents the width of the rectangle.
      - 'this.height' is an instance attribute that represents the height of the rectangle.

      The constructor initializes the 'width' and 'height' attributes with the values
      passed as arguments 'w' and 'h' when creating a new Rectangle instance.
    */

    /*
      Check if 'w' and 'h' are positive integers.
      If either is not a positive integer or if either is 0, create an empty object.
      Otherwise, initialize the 'width' and 'height' attributes.
    */
    if (Number.isInteger(w) && w > 0 && Number.isInteger(h) && h > 0) {
      this.width = w;
      this.height = h;
    } else {
      /* Create an empty object if conditions are not met */
      this.width = undefined;
      this.height = undefined;
    }
  }
}
module.exports = Rectangle;
